File: leetcode-submissions/515_Find_Largest_Value_in_Each_Tree_Row.ts
function largestValues(root: TreeNode | null): number[] {
    if (root === null) return []; // If the tree is empty, return an empty array

    const queue: TreeNode[] = [root]; // Initialize the queue with the root node
    const ans: number[] = []; // To store the largest values

    while (queue.length > 0) {
        let currSize: number = queue.length;
        const temp: number[] = []; // To store values at the current level

        while (currSize-- > 0) {
            const CurrNode = queue.shift(); // Remove the node from the queue

            if (CurrNode) { // Ensure the node is not null
                temp.push(CurrNode.val); // Push the value of the current node
                if (CurrNode.left) {
                    queue.push(CurrNode.left); // Add left child to the queue
                }
                if (CurrNode.right) {
                    queue.push(CurrNode.right); // Add right child to the queue
                }
            }
        }
        ans.push(Math.max(...temp)); // Add the largest value at the current level
    }

    return ans;
}

File: leetcode-submissions/515_Find_Largest_Value_in_Each_Tree_Row.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> largestValues(TreeNode* root) {
        vector<int>ans ;
        queue<TreeNode*>q;
        q.push(root);
        while(not q.empty()){
            vector<int>temp;

            int currSize=q.size();
            while(currSize--){
                TreeNode * currNode =q.front();
                q.pop();
                temp.push_back(currNode->val);
                if(currNode->left){
                    q.push(currNode->left);
                }
                if(currNode->right){
                    q.push(currNode->right);


                }

                
            }
            int maxEle=*max_element(temp.begin(),temp.end());
            ans.push_back(maxEle);
        }
        return ans ;
    }
};